Segment Tree/LazyPropagation/README.markdown

```diff
@@ -0,0 +1,308 @@
+# Lazy Propagation in Segment Tree
+
+In previous implement about the Segment Tree by **Artur Antonov**, it's a strong data structure with *Generic* `<T>`. And we can pass a closure parameter `function: (T, T) -> T` to reflect the relationship between parent and child node. And in particular, Generic can solve multiple strings stitching problem. It's just like the sample in Playground:
+
+```swift
+stringSegmentTree.replaceItem(at: 0, withItem: "I")
+stringSegmentTree.replaceItem(at: 1, withItem: " like")
+stringSegmentTree.replaceItem(at: 2, withItem: " algorithms")
+stringSegmentTree.replaceItem(at: 3, withItem: " and")
+stringSegmentTree.replaceItem(at: 4, withItem: " swift")
+stringSegmentTree.replaceItem(at: 5, withItem: "!")
+print(stringSegmentTree.query(leftBound: 0, rightBound: 5))
+// "I like algorithms and swift!"
+```
+
+The use of `<T>` is so exciting. But we seldom use the Segment Tree to solve string problem instead of *Suffix Array*. And the Segment Tree is a kind of *Interval Tree* to solve the Interval Problem in mathemtics and statistics, which is a structure for storing intervals, or segments, and allows querying which of the stored segments contain a given point. A segment tree for a set *I* of n intervals uses `O(nlogn)` storage and can be built in `O(nlogn)` time. Segment trees support searching for all the intervals that contain a query point in O(log n+k), k being the number of retrieved intervals or segments.
+
+But that is common Segment Tree. By **Lazy Propagation**, we can implement to modify an interval in `O(logn)` time. Let's explore together in following:
+
+## `PushUp` - update to the top
+
+At first, we reference the implement of **Artur Antonov** about Segment Tree. This code contained *build*, *single update* and *interval query* three operation. The implement of *build* and *single update* operation is following:
+
+```swift
+public init(array: [T], leftBound: Int, rightBound: Int, function: @escaping (T, T) -> T) {
+	self.leftBound = leftBound
+	self.rightBound = rightBound
+	self.function = function
+	// ①
+	if leftBound == rightBound {
+		value = array[leftBound]
+	} 
+	// ②
+	else {
+		let middle = (leftBound + rightBound) / 2
```

Comment:
update
```swift
let middle = leftBound + (rightBound - leftBound) / 2
```

Reply:
@remlostime 
After my thought, I think it shouldn't change the process of evaluate `middle`. It should be `(leftBound + rightBound) / 2`. Because they are `integer` or not `float`.

There is a data testing for you. 

![screen shot 2017-09-18 at 11 28 51](https://user-images.githubusercontent.com/7804535/30528332-a42e1754-9c64-11e7-8ac9-4cc878956775.png)

Reply:
Try this one `l = Int.Max - 1`, `r = Int.Max`, what will happen?
FYI - Some interesting article talking about this overflow issue
http://www.geeksforgeeks.org/binary-search/
https://stackoverflow.com/a/6735322
https://research.googleblog.com/2006/06/extra-extra-read-all-about-it-nearly.html
https://stackoverflow.com/questions/21101110/calculating-midpoint-index-in-binary-search

Reply:
Thanks for your articles. And they proved the `l + (r - l) / 2` is feasible (when the `r > l`)! 😁  And I will fix it.

--- 

And I also want to add my opinions for this problem. 

In memory storage, we can't store a array whose length is `Int.max` or `Int.max / 2`. **So the result of `l + r` must be low than `Int.max`.** 

![image](https://user-images.githubusercontent.com/7804535/30635544-526807f0-9e25-11e7-8431-b8b8ff377e39.png)

And we use `Segment Tree` to store a linear array, and we need to memset a bigger space in memory, which probably needs `O(2^(n/2) + 1)` storage from `O(n)`. If we wanted to save more space, we need to try the **zkw Segment Tree** or to **discretize data** by a hash table.  : )

And then, I'll try to implement the discretize data for the Segment Tree as an extension. 

In the end, thank you again.  🤝

Reply:
Sure thing! :)